engine/src/cmd/collision.cpp

```diff
@@ -0,0 +1,263 @@
+
+static float front_collision_hack_angle = cos( 3.1415926536f * GameConfig::GetVariable( "physics", "front_collision_avoidance_hack_angle", 40)/180.0f );
+
+static const bool collision_damage_to_ai = GameConfig::GetVariable( "physics", "collisionDamageToAI", false);
+
+static const bool crash_dock_unit = GameConfig::GetVariable( "physics", "unit_collision_docks", false);
+
+static const bool crash_dock_hangar = GameConfig::GetVariable( "physics", "only_hangar_collision_docks", false);
+
+// Disabled bouncing missile option. Missiles always explode when colliding with something.
+//static const bool does_missile_bounce = GameConfig::GetVariable( "physics", "missile_bounce", false);
+
+
+
+Collision::Collision(Unit* unit, const QVector& location, const Vector& normal):
+    unit(unit), location(location), normal(normal)
+{
+    cockpit = _Universe->isPlayerStarship( unit ); // smcp/thcp
+    unit_type = unit->isUnit();
+    is_player_ship = _Universe->isPlayerStarship(unit);
+    mass = std::max(unit->GetMass(), minimum_mass);
+    position = unit->Position();
+    velocity = unit->GetVelocity();
+}
+
+
+
+
+// This function handles the initial reaction of the unit to hitting the other unit
+// Note: I'm changing the expected behavior here
+// Return value indicates whether to continue processing
+/*UNITPTR,
+PLANETPTR,
+BUILDINGPTR,
+NEBULAPTR,
+ASTEROIDPTR,
+ENHANCEMENTPTR,
+MISSILEPTR*/
+void Collision::shouldApplyForceAndDealDamage(Unit* other_unit)
+{
+    // Collision with a nebula does nothing
+    if(other_unit->isUnit() == NEBULAPTR)
+    {
+        return;
+    }
+
+    // Collision with a enhancement improves your shield apparently
+    if(other_unit->isUnit() == ENHANCEMENTPTR)
+    {
+        apply_force = true;
+        return;
+    }
+
+
+    switch(unit_type)
+    {
+    // Missiles and asteroids always explode on impact with anything except Nebula and Enhancement.
+    case MISSILEPTR:
+        // Missile should explode when killed
+        // If not, uncomment this
+        //((Missile*)unit)->Discharge();
+        unit->Kill();
+        return;
+
+    case ASTEROIDPTR:
+        unit->Kill();
+        return;
+
+    // Planets and Nebulas can't be killed right now
+    case PLANETPTR:
+    case NEBULAPTR:
+        return;
+
+    // Buildings should not calculate actual damage
+    case BUILDINGPTR:
+        return;
+
+    // Units (ships) should calculate actual damage
+    case UNITPTR:
+        apply_force = true;
+        deal_damage = true;
+        return;
+
+    // Not sure what an enhancement is, but it looks like it's something that can increase the shields of the unit it collided with.
+    case ENHANCEMENTPTR:
+        if (other_unit->isUnit() == ASTEROIDPTR)
+        {
+            apply_force = true;
+            return;
+        }
+
+        double percent;
+        char tempdata[sizeof (unit->shield)];
+        memcpy( tempdata, &unit->shield, sizeof (unit->shield) );
```

Comment:
This should use sizeof(type) instead of sizeof(var)

Reply:
Done. Also, this is old code and I need to balance width with depth here. I could spend the next year polishing a few files.